conjure-java-core/src/main/java/com/palantir/conjure/java/types/SafetyEvaluator.java

```diff
@@ -0,0 +1,285 @@
+        this.definitionMap = definitionMap;
+    }
+
+    public Optional<LogSafety> evaluate(TypeDefinition def) {
+        return Preconditions.checkNotNull(def, "TypeDefinition is required")
+                .accept(new TypeDefinitionSafetyVisitor(definitionMap, new HashSet<>()));
+    }
+
+    public Optional<LogSafety> evaluate(Type type) {
+        return Preconditions.checkNotNull(type, "TypeDefinition is required")
+                .accept(new TypeDefinitionSafetyVisitor(definitionMap, new HashSet<>()).fieldVisitor);
+    }
+
+    public Optional<LogSafety> evaluate(Type type, Optional<LogSafety> declaredSafety) {
+        return declaredSafety.or(() -> evaluate(type));
+    }
+
+    private static final class TypeDefinitionSafetyVisitor implements TypeDefinition.Visitor<Optional<LogSafety>> {
+        private final Set<TypeName> inProgress;
+        private final Type.Visitor<Optional<LogSafety>> fieldVisitor;
+
+        private TypeDefinitionSafetyVisitor(Map<TypeName, TypeDefinition> definitionMap, Set<TypeName> inProgress) {
+            this.inProgress = inProgress;
+            this.fieldVisitor = new FieldSafetyVisitor(definitionMap, this);
+        }
+
+        @Override
+        public Optional<LogSafety> visitAlias(AliasDefinition value) {
+            return with(value.getTypeName(), () -> getSafety(value.getAlias(), value.getSafety()));
+        }
+
+        @Override
+        public Optional<LogSafety> visitEnum(EnumDefinition _value) {
+            return ENUM_SAFETY;
+        }
+
+        @Override
+        public Optional<LogSafety> visitObject(ObjectDefinition value) {
+            return with(value.getTypeName(), () -> {
+                Optional<LogSafety> safety = Optional.of(LogSafety.SAFE);
+                for (FieldDefinition field : value.getFields()) {
+                    safety = combine(safety, getSafety(field.getType(), field.getSafety()));
+                }
+                return safety;
+            });
+        }
+
+        @Override
+        public Optional<LogSafety> visitUnion(UnionDefinition value) {
+            return with(value.getTypeName(), () -> {
+                Optional<LogSafety> safety = UNKNOWN_UNION_VARINT_SAFETY;
+                for (FieldDefinition variant : value.getUnion()) {
+                    safety = combine(safety, getSafety(variant.getType(), variant.getSafety()));
+                }
+                return safety;
+            });
+        }
+
+        @Override
+        public Optional<LogSafety> visitUnknown(String unknownType) {
+            throw new IllegalStateException("Unknown type: " + unknownType);
+        }
+
+        private Optional<LogSafety> with(TypeName typeName, Supplier<Optional<LogSafety>> task) {
+            if (!inProgress.add(typeName)) {
+                // Given recursive evaluation, we return the least restrictive type: SAFE.
+                return Optional.of(LogSafety.SAFE);
```

Comment:
Not sure I follow here. Why is the inProgress tracking needed here? I don't see the case yet, where this where we start processing a type name that is already in progress?

Reply:
That's correct, I may define something along these lines (although in practice they tend to have a great deal more indirection)

```yaml
MyNode:
  fields:
    foo:
      type: integer
      safety: unsafe
    next: optional<MyNode>
```

We need to avoid infinite recursion in this case

Reply:
Ah, I see. Because `task.get()` will evaluate the whole subtree which might recurse. Thanks!